Zend/zend_dimension_handlers.h

```diff
@@ -0,0 +1,39 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Author: Gina Peter Banyard <girgias@php.net>                         |
+   +----------------------------------------------------------------------+
+*/
+
+/* This really should not be needed... */
+#ifndef ZEND_DIMENSION_HANDLERS_H
+#define ZEND_DIMENSION_HANDLERS_H
+
+//BEGIN_EXTERN_C()
+
+#include "zend_types.h"
+
+typedef struct _zend_class_dimensions_functions {
+	/* rv is a slot provided by the callee that is returned */
+	zval *(*/* const */ read_dimension)(zend_object *object, zval *offset, zval *rv);
+	bool  (*/* const */ has_dimension)(zend_object *object, zval *offset);
+	zval *(*/* const */ fetch_dimension)(zend_object *object, zval *offset, zval *rv);
```

Comment:
I've looked at a few places and didn't see any. Does the engine ever call `has_dimension` without ever calling one of the other handlers? It seems like it's always followed by one.

Potentially, we should stop calling `has_dimension` and rely on the error handling that still seems to exis in the other dimension handlers. It certainly looks like they cannot assume that they are only called when `has_dimension` is true. And in the case of hash lookups, the lookup is going to get repeated if we call `has_dimension` and `read_dimension`, for example.

Reply:
I don't think it's reasonable to "just" call the `read_dimension` and not call the `had_dimension` handler, as this would be confusing for users.

I suppose what I could do is what ArrayAccess used to do is cache the `zend_function` pointers so we don't need to look them up.

Reply:
> I don't think it's reasonable to "just" call the read_dimension and not call the had_dimension handler, as this would be confusing for users.

Can you expand? What's confusing?

Reply:
I don't mean userland, I mean internally. There are C-level calls to `has_dimension` which get followed up by calls to things like `read_dimension`. Note that in your example the hash lookup happens twice. Internal calls should be able to avoid that, I think. Or at least, I would try to design the APIs that way. If `read_dimension` returns `NULL`, it clearly did not have that dimension. See what I'm saying?

Reply:
The difference between a C handler and the userland function is basically non-existent, having it be different is just confusing.

The hash look_up would have only happened if you were calling the userland version _anyway_  and did not have an optimized C version of the handler.

The API is literally just `has_dimension` == call to `offsetExists()`, `read_dimension` == call to `offsetGet()`, etc.
But all the internal extensions basically implement an overloaded C version which does not call a userland function.

Also about the hash look-up, this should be a non-issue now as I decided to use a union to either store the `zend_function*` for userland classes or the C handler function for internal extensions.